workflows/auto-pr: be more forceful

woodruffw opened the pull request:
~~WIP, still working on debugging the `auto-pr` workflow.~~

woodruffw commented:
Successful run: https://github.com/Homebrew/brew-pip-audit/actions/runs/4091289431/jobs/7055271713

xref https://github.com/Homebrew/homebrew-core/pull/122291

woodruffw commented:
Some other things we need to do here (H/T @chenrui333):

1. Bump the revision of each formulae
2. Provide a better summary of the update in the PR description (which formulae are being updated + which vulnerabilities are being patched by ID)

woodruffw commented:
> * Bump the revision of each formulae

There's no existing API for this (only a `dev-cmd`), so I'll probably need to break one out.

alex commented:
We don't need to bump revisions, that's no longer done I was told.

alex commented on generate-prs.rb:
```diff
@@ -2,16 +2,14 @@
-SKIP_FORMULA = []
-# TODO: remove once we're confident this works
-ONLY_FORMULA = ["grokmirror"]
+SKIP_FORMULA = ["grokmirror"]
```
Why is this in skip?

woodruffw replied:
Just me experimenting -- the initial `grokmirror` PR was merged but the corresponding file wasn't deleted here yet (maybe this workflow needs to do that?), so I didn't want any confused state.

alex replied:
It won't get deleted until the other action runs, it's what cleans these up.

Even if this tries to upgrade, it won't send a PR when it sees one is already open though.

woodruffw replied:
> Even if this tries to upgrade, it won't send a PR when it sees one is already open though.

Right, forgot about that...

woodruffw commented:
> We don't need to bump revisions, that's no longer done I was told.

I thought so as well, but @p-linnane told me that we still need `revision` bumping -- without it these changes will only affect future installers of the same version of a formula, whereas with a `revision` bump it'll be correctly handled by auto-upgrades, etc.

Ref: https://docs.brew.sh/Formula-Cookbook#formulae-revisions

p-linnane commented:
If you want to make changes to a formula that users will receive when running `brew upgrade`, but aren't incrementing the version, you will need a `revision`. The most common case for this is rebuilding dependents against an updated library.

alex commented:
Then we do want to bump version. Maybe we should get this working end-to-end (and automatically) and then cycle back to `revision`, since we'll probably want to add that into the existing library.

woodruffw commented:
> Maybe we should get this working end-to-end (and automatically) and then cycle back to `revision`, since we'll probably want to add that into the existing library.

Agreed; I'll shore this up and get it merged.

woodruffw commented:
Got a little further, but now we explode if a bump doesn't happen to produce any changes (I would expect this to never happen, so I need to debug it):

```
Updating resources for termtosvg
Switched to a new branch 'bump-python-resources-termtosvg'
On branch bump-python-resources-termtosvg
nothing to commit, working tree clean
/usr/local/Homebrew/Library/Homebrew/utils.rb:324:in `safe_system': Failure while executing; `git commit --no-edit --verbose --message=termtosvg:\ bump\ python\ resources -- /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/termtosvg.rb` exited with 1. (ErrorDuringExecution)
	from /usr/local/Homebrew/Library/Homebrew/utils/github.rb:607:in `block in create_bump_pr'
	from /usr/local/Homebrew/Library/Homebrew/extend/pathname.rb:303:in `block in cd'
	from /usr/local/Homebrew/Library/Homebrew/extend/pathname.rb:303:in `chdir'
	from /usr/local/Homebrew/Library/Homebrew/extend/pathname.rb:303:in `cd'
	from /usr/local/Homebrew/Library/Homebrew/utils/github.rb:564:in `create_bump_pr'
	from generate-prs.rb:41:in `block in <main>'
	from generate-prs.rb:7:in `each'
	from generate-prs.rb:7:in `<main>'
```

woodruffw commented:
Some more context:

```
Updating resources for termtosvg
Switched to a new branch 'bump-python-resources-termtosvg'
M	Formula/detect-secrets.rb
M	Formula/fdroidserver.rb
M	Formula/onlykey-agent.rb
On branch bump-python-resources-termtosvg
Changes not staged for commit:
	modified:   detect-secrets.rb
	modified:   fdroidserver.rb
	modified:   onlykey-agent.rb

no changes added to commit
/usr/local/Homebrew/Library/Homebrew/utils.rb:324:in `safe_system': Failure while executing; `git commit --no-edit --verbose --message=termtosvg:\ bump\ python\ resources -- /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/termtosvg.rb` exited with 1. (ErrorDuringExecution)
```

...so, separately, we should probably do some additional cleanup to ensure each branch is clean before bumping.

woodruffw commented:
Feedback:

1. We should skip deprecated and disabled formulae (thanks @iMichka)
2. It looks like sometimes we downgrade versions (e.g., <https://github.com/Homebrew/homebrew-core/pull/122410/files>), probably because someone manually upgraded them only in Homebrew (which they shouldn't have)

woodruffw commented:
65552ae skips deprecated/disabled formulae.

woodruffw commented:
Another thing: it looks like we're not updating Python resources on `HEAD`. I'll look into it.

woodruffw commented:
And another: we should think about only updating the listed vulnerable subdependencies, not all out-of-date subdependencies. This would make the diffs smaller and more topical, and would avoid any unnecessary breakage.

woodruffw commented:
Another nice to have: the PR message body should probably include a URL back to the CI job, for rapid debugging.

alex commented on formula2requirements.rb:
```diff
@@ -16,6 +16,14 @@
 
+  # Skip deprecated and disabled formulae but not before removing any
+  # previously generated requirements, if present.
+  if f.deprecated? || f.disabled?
```
deprecated formula are still installable right? Why wouldn't we want to submit vuln fixes for them.

woodruffw replied:
This came up during IRL review: deprecated formulae don't receive updates, meaning that the bottle CI skips them. So any updates we perform to dependencies on deprecated formulae wouldn't reach users (and aren't even tested with `brew test`, since the CI just ignores them).

IMO most of the deprecated formulae identified here are good candidates for disablement/removal anyways, so I'll look into that.

woodruffw commented:
This is mostly working, so I'm pulling the trigger on it.
